### DISCUSSION TITLE
Change a select clause + add a join automatically

### DISCUSSION THREAD

**Discussion by shreyanshgautam24** [sqlalchemy/sqlalchemy#9641] [Usage Questions]
Hello,

I have this schema:

```
class Question(Base):
    __tablename__ = "question"
    idn = Column(Integer, primary_key=True)
    text = Column("text", String)

    translations = relationship("Translation", backref="question")

class Translation(Base):
    __tablename__ = "translation"
    idn = Column(Integer, primary_key=True)
    qid = Column(Integer, ForeignKey("question.idn"))
    lang = Column(String)
    text = Column(String)
```

I want to automatically join Question with a single Translation (lang filter) when the real query happens. Language will be supplied runtime. I tried to think of this as a property, but that doesn't seem to work.

Currently, I am using a  simple function to achieve this:

```
def translate(query, lang):
    cols = list(query.selectable.selected_columns)
    index = next(
        (
            i
            for i, c in enumerate(cols)
            if [c.name](http://c.name/) == "text" and [c.table.name](http://c.table.name/) == "question"
        ),
        None,
    )
    text = case(
        [(Translation.text.is_not(None), Translation.text)],
        else_=Question.text,
    )
    cols[index] = text.label("text")
    query._set_entities(cols)  # noqa:W0212
    query = query.outerjoin(
        Translation,
        and_(Question.idn == Translation.qid, Translation.lang == lang),
    )
    return query
```

Usage: `query = translate(query, "it")`

1. Is there a good place to hook this so that it's baked into the Question table? The tricky part is when and where to pass "it". I guess that can be only done during session.query. If so, how to do that? (I know about bind-params, so I am thinking that will be an option here.)

2. Is there a better approach to solve this problem? I tried looking up examples, but didn't find one that fits my purpose (should be simple, should be at a single place etc.).

SQLAlchemy 1.3

**Comment by zzzeek** [comment]
there's no ".selected_columns" attribute in SQLAlchemy 1.3.  can you double check what version you are on?  because the approach to this kind of problem changes as of SQLAlchemy 1.4.    Doing this in SQLAlchemy 1.3 would not be as easy.

**Reply by shreyanshgautam24** [reply]
Sorry, that is correct. This code was tried on 1.4; however, we are trying to make it work on 1.3. Are there no equivalent ways to to do this in 1.3? If not, if we upgrade to 1.4, is this an OK solution, or is there a better way to achieve this?

**Reply by zzzeek** [reply]
1.4 has an all-encompassing hook that you can intercept all ORM statements called [do_orm_execute](https://docs.sqlalchemy.org/en/14/orm/events.html#sqlalchemy.orm.SessionEvents.do_orm_execute) , where you'd be passed a core `select()` construct in all cases (even if the original operation used `session.query()`).  The `select()` then has all the fairly rich manipulation methods you want to look for.

in 1.3, you'd use the [before_compile](https://docs.sqlalchemy.org/en/13/orm/events.html#sqlalchemy.orm.events.QueryEvents.before_compile) hook.  totally different, and this hook will not be very effective in 1.4 due to the caching system introduction.
